src/ray/gcs/gcs_server/gcs_resource_report_poller.h

```diff
@@ -0,0 +1,133 @@
+                 std::function<void(const Status &,
+                                    const rpc::RequestResourceReportReply &)>
+                     callback) {
+                auto raylet_client = raylet_client_pool->GetOrConnectByAddress(address);
+                raylet_client->RequestResourceReport(callback);
+              });
+
+  ~GcsResourceReportPoller();
+
+  /// Start a thread to poll for resource updates.
+  void Start();
+
+  /// Stop polling for resource updates.
+  void Stop();
+
+  /// Event handler when a new node joins the cluster.
+  void HandleNodeAdded(const std::shared_ptr<rpc::GcsNodeInfo> &node_info)
+      LOCKS_EXCLUDED(mutex_);
+
+  /// Event handler when a node leaves the cluster.
+  void HandleNodeRemoved(const std::shared_ptr<rpc::GcsNodeInfo> &node_info)
+      LOCKS_EXCLUDED(mutex_);
+
+ private:
+  // An asio service which does the polling work.
+  boost::asio::io_context polling_service_;
+  // The associated thread it runs on.
+  std::unique_ptr<std::thread> polling_thread_;
+  // Timer tick to check when we should do another poll.
+  PeriodicalRunner ticker_;
+
+  // The maximum number of pulls that can occur at once.
+  const uint64_t max_concurrent_pulls_;
+  // The number of ongoing pulls.
+  uint64_t inflight_pulls_;
+  // The resource manager which maintains GCS's view of the cluster's resource
+  // utilization.
+  std::shared_ptr<GcsResourceManager> gcs_resource_manager_;
+  // The shared, thread safe pool of raylet clients, which we use to minimize connections.
+  std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool_;
+  // Handle receiving a resource report (e.g. update the resource manager).
+  std::function<void(const rpc::ResourcesData &)> handle_resource_report_;
+
+  // Return the current time in miliseconds
+  std::function<int64_t(void)> get_current_time_milli_;
+  // Send the `RequestResourceReport` RPC.
+  std::function<void(
+      const rpc::Address &, std::shared_ptr<rpc::NodeManagerClientPool> &,
+      std::function<void(const Status &, const rpc::RequestResourceReportReply &)>)>
+      request_report_;
+  // The minimum delay between two pull requests to the same thread.
+  int64_t poll_period_ms_;
+
+  struct PullState {
+    NodeID node_id;
+    rpc::Address address;
+    int64_t last_pull_time;
+    int64_t next_pull_time;
+
+    ~PullState() {}
+  };
+
+  // A global lock for internal operations. This lock is shared between the main thread
+  // and polling thread, so we should be mindful about how long we hold it.
+  absl::Mutex mutex_;
+  // All the state regarding how to and when to send a new pull request to a raylet.
+  std::unordered_map<NodeID, std::shared_ptr<PullState>> nodes_;
+  // The set of all nodes which we are allowed to pull from. We can't necessarily pull
+  // from this list immediately because we limit the number of concurrent pulls.
+  std::deque<std::shared_ptr<PullState>> to_pull_queue_;
```

Comment:
Specify in the comment that these should be ordered by the next timestamp to pull.